### PR TITLE
refactor: use rocksdb_wrapper::write_batch_put to reimplement check_and_set

### PR DESCRIPTION
**Pull request by levy5307**
### What problem does this PR solve? 
use rocksdb_wrapper::write_batch_put to reimplement check_and_set

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
```
>>> use temp
OK
>>> use temp
OK
>>> check_and_set a -c b -t exist -s b1 -v c1 -r
hash_key: "a"
check_sort_key: "b"
check_type: exist
set_sort_key: "b1"
set_value: "c1"
set_value_ttl_seconds: 0
return_check_value: true

Set failed, because check not passed.

Check value not exist.

app_id          : 2
partition_index : 4
decree          : 9
server          : 10.232.55.210:34802
>>> set a b c
OK

app_id          : 2
partition_index : 4
decree          : 10
server          : 10.232.55.210:34802
>>> get a b 
"c"

app_id          : 2
partition_index : 4
server          : 10.232.55.210:34802
>>> check_and_set a -c b -t exist -s b1 -v c1 -r
hash_key: "a"
check_sort_key: "b"
check_type: exist
set_sort_key: "b1"
set_value: "c1"
set_value_ttl_seconds: 0
return_check_value: true

Set succeed.

Check value: "c"

app_id          : 2
partition_index : 4
decree          : 12
server          : 10.232.55.210:34802
>>> get a b1
"c1"

app_id          : 2
partition_index : 4
server          : 10.232.55.210:34802
>>> set a b c 1
OK

app_id          : 2
partition_index : 4
decree          : 14
server          : 10.232.55.210:34802
>>> check_and_set a -c b -t exist -s b2 -v c2 -r
hash_key: "a"
check_sort_key: "b"
check_type: exist
set_sort_key: "b2"
set_value: "c2"
set_value_ttl_seconds: 0
return_check_value: true

Set failed, because check not passed.

Check value not exist.

app_id          : 2
partition_index : 4
decree          : 15
server          : 10.232.55.210:34802
>>> get a b2
Not found

app_id          : 2
partition_index : 4
server          : 10.232.55.210:34802
>>> check_and_set a -c b -t not_exist -s b2 -v c2 -r
hash_key: "a"
check_sort_key: "b"
check_type: not_exist
set_sort_key: "b2"
set_value: "c2"
set_value_ttl_seconds: 0
return_check_value: true

Set succeed.

Check value not exist.

app_id          : 2
partition_index : 4
decree          : 17
server          : 10.232.55.210:34802
>>> get a b2
"c2"

app_id          : 2
partition_index : 4
server          : 10.232.55.210:34802
```

### CODE REVIEW

File: src/server/pegasus_write_service_impl.h [hunk 1]
```diff
@@ -338,22 +338,24 @@ class pegasus_write_service::impl : public dsn::replication::replica_base
...
+                static_cast<uint32_t>(update.set_expire_ts_seconds));
         } else {
             // check not passed, write empty record to update rocksdb's last flushed decree
-            resp.error = db_write_batch_put(decree, dsn::string_view(), dsn::string_view(), 0);
+            resp.error = _rocksdb_wrapper->write_batch_put(
+                decree, dsn::string_view(), dsn::string_view(), 0);
         }
         if (resp.error) {
-            clear_up_batch_states(decree, resp.error);
+            _rocksdb_wrapper->clear_up_write_batch();
```
Comment (acelyc111) [inline src/server/pegasus_write_service_impl.h:352]: Use an RAII way to optimize `clear_up_write_batch `.

File: src/server/rocksdb_wrapper.cpp [hunk 2]
```diff
@@ -66,5 +75,109 @@ int rocksdb_wrapper::get(dsn::string_view raw_key, /*out*/ db_get_context *ctx)
...
+                       utils::c_escape_string(hash_key),
+                       utils::c_escape_string(sort_key),
+                       expire_sec);
+    }
+    return s.code();
+}
+
+int rocksdb_wrapper::write(int64_t decree)
+{
+    dassert(_write_batch->Count() != 0, "");
```
Comment (acelyc111) [inline src/server/rocksdb_wrapper.cpp:142]: Fill the error message.

File: src/server/rocksdb_write_batch_cleaner.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/*
```
Comment (acelyc111) [inline src/server/rocksdb_write_batch_cleaner.h:1]: You can use `dsn::defer`, not need to reimplement it.

File: src/server/pegasus_write_service_impl.h [hunk 4]
```diff
@@ -349,13 +350,13 @@ class pegasus_write_service::impl : public dsn::replication::replica_base
                 decree, dsn::string_view(), dsn::string_view(), 0);
         }
         if (resp.error) {
-            _rocksdb_wrapper->clear_up_write_batch();
+            rocksdb_write_batch_cleaner cleaner(_rocksdb_wrapper.get());
```
Comment (acelyc111) [inline src/server/pegasus_write_service_impl.h:353]: Define it in the function body scope, then you can only define it once.

Comment (levy5307) [reply src/server/pegasus_write_service_impl.h]: OK, I got what you mean.

**Review by acelyc111 — DISMISSED** [review]

File: src/server/rocksdb_wrapper.h [hunk 5]
```diff
@@ -46,11 +53,37 @@ class rocksdb_wrapper : public dsn::replication::replica_base
...
+
 private:
+    uint32_t db_expire_ts(uint32_t expire_ts);
+
     rocksdb::DB *_db;
     rocksdb::ReadOptions &_rd_opts;
+    std::unique_ptr<pegasus_value_generator> _value_generator;
+    std::unique_ptr<rocksdb::WriteBatch> _write_batch;
+    std::unique_ptr<rocksdb::WriteOptions> _wt_opts;
+    rocksdb::ColumnFamilyHandle *_meta_cf;
```
Comment (neverchanje) [inline src/server/rocksdb_wrapper.h:77]:
```suggestion
    pegasus_value_generator _value_generator;
    rocksdb::WriteBatch _write_batch;
    rocksdb::WriteOptions _wt_opts;
    rocksdb::ColumnFamilyHandle *_meta_cf;
```

Why not use value?

Comment (levy5307) [reply src/server/rocksdb_wrapper.h:77]: To reduce file dependencies

**Review by neverchanje — APPROVED** [review]

**Review by hycdong — APPROVED** [review]

**Pull request merged**